### PR TITLE
All settings are now stored in config.php

### PR DESCRIPTION
**Pull request by ArthurHoaro**
options.php is not supported anymore, but its content will be automatically saved into config.php

Fixes shaarli/Shaarli#41

*TODO*: update [documentation](https://github.com/shaarli/Shaarli/wiki#configuration).

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -2373,6 +2375,14 @@ function invalidateCaches()
     pageCache::purgeCache();   // Purge page cache shared by sessions.
 }
 
+// Milestone 0.9 - shaarli/Shaarli#41: options.php is not supported anymore
```
Comment (virtualtam) [inline index.php]: could this snippet be encapsulated in a `cleanupConfig()` method?

File: index.php [hunk 2]
```diff
@@ -2128,17 +2125,22 @@ function json_encode($data) {
...
-    $config .= '$GLOBALS[\'titleLink\']='.var_export($GLOBALS['titleLink'],true).'; ';
-    $config .= '$GLOBALS[\'redirector\']='.var_export($GLOBALS['redirector'],true).'; ';
-    $config .= '$GLOBALS[\'disablesessionprotection\']='.var_export($GLOBALS['disablesessionprotection'],true).'; ';
-    $config .= '$GLOBALS[\'disablejquery\']='.var_export($GLOBALS['disablejquery'],true).'; ';
-    $config .= '$GLOBALS[\'privateLinkByDefault\']='.var_export($GLOBALS['privateLinkByDefault'],true).'; ';
-    $config .= '$GLOBALS[\'config\'][\'ENABLE_RSS_PERMALINKS\']='.var_export($GLOBALS['config']['ENABLE_RSS_PERMALINKS'], true).'; ';
-    $config .= '$GLOBALS[\'config\'][\'ENABLE_UPDATECHECK\']='.var_export($GLOBALS['config']['ENABLE_UPDATECHECK'], true).'; ';
-    $config .= '$GLOBALS[\'config\'][\'HIDE_PUBLIC_LINKS\']='.var_export($GLOBALS['config']['HIDE_PUBLIC_LINKS'], true).'; ';
-    $config .= ' ?>';
+    $config='<?php '. PHP_EOL;
```
Comment (virtualtam) [inline index.php]: could both sections (`writeConfig()`, and the removal of `options.php`) be moved under a new `application/Config.php` module?

**Comment by ArthurHoaro** [comment]
@virtualtam 
> could both sections (writeConfig(), and the removal of options.php) be moved under a new application/Config.php module?

Absolutely.

**Comment by ArthurHoaro** [comment]
Ok it seems to be working just fine. I'll let you review it.

File: application/Config.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
+<?php
+/**
+ * Functions related to configuration management.
+ */
+
+// These fields are required in configuration.
+$MANDATORY_FIELDS = ['login', 'hash', 'salt', 'timezone', 'title', 'titleLink', 'redirector', 'disablesessionprotection', 'privateLinkByDefault'];
```
Comment (virtualtam) [inline application/Config.php]: don't hesitate to spread long lists over several lines :)

File: application/Config.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * @param $config: array containing all configuration fields.
+ * @param $isLoggedIn: boolean.
+ *
+ * @throws MissingFieldConfigException: a mandatory field has not been provided in $config.
+ * @throws UnauthorizedConfigException: user is not authorize to change configuration.
+ * @throws Exception: an error occured while writing the new config file.
+ */
+function writeConfig($config, $isLoggedIn)
+{
+    global $MANDATORY_FIELDS;
```
Comment (virtualtam) [inline application/Config.php]: `$MANDATORY_FIELDS` could be injected as a parameter (this avoids access to external stuff)

File: application/Config.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        }
+        $config['config']['CONFIG_FILE'] = $config_file;
+        writeConfig($config, $isLoggedIn);
+
+        unlink($config['config']['DATADIR'].'/options.php');
+    }
+}
+
+/**
+ * Exception used if a mandatory field in missing in configuration.
```
Comment (virtualtam) [inline application/Config.php]: is

File: index.php [hunk 6]
```diff
@@ -69,6 +66,7 @@
 // Shaarli library
 require_once 'application/LinkDB.php';
 require_once 'application/Utils.php';
+require_once 'application/Config.php';
```
Comment (virtualtam) [inline index.php]: should imports be sorted?

when working on #71, I looked at PHP's recommendations for autoloading classes, but it seemed to involve a bit too much "magic" (dynamic loading, by infering the file to include from a class' name)

File: index.php [hunk 7]
```diff
@@ -2373,6 +2367,12 @@ function invalidateCaches()
     pageCache::purgeCache();   // Purge page cache shared by sessions.
 }
 
+try {
+    mergeDeprecatedConfig($GLOBALS, isLoggedIn());
+} catch(Exception $e) {
+    // don't care
```
Comment (virtualtam) [inline index.php]: see PHP's `error_log()` function:
- if specified, it logs messages in a file specified by the server configuration (vhost, htaccess...), which is useful for debugging
- it's non-blocking

File: tests/ConfigTest.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
+<?php
+/**
+ * Config' tests
```
Comment (virtualtam) [inline tests/ConfigTest.php]: unit tests <3

**Comment by virtualtam** [comment]
- works fine
- tests run smoothly (thanks for adding them :) )
- usage of Exceptions to handle errors is elegant, maybe add some calls to PHP logging functionalities
- +1 for the TODOs to remove JS error handling & display
- for code that's been moved/added to separate files, it's easy to run static analysis tools: `make code_sniffer`, `make mess_detector`; though we still have to define which coding rules to stick to, it helps keeping code homogenous (indents, spacing...)

File: application/Config.php (hunk 4 above)
Comment (ArthurHoaro) [inline application/Config.php]: `$MANDATORY_FIELDS` is here used as a constant. Doesn't make a lot of sense to me to pass it as a parameter.

File: index.php (hunk 6 above)
Comment (ArthurHoaro) [inline index.php]: Don't know... maybe we can think of autoloading, namespaces, etc. But yet there is no interdependency between files, so it doesn't really matter.

**Comment by ArthurHoaro** [comment]
Thanks for your code review.

For coding rules, is there an issue where this is discussed?

**Comment by ArthurHoaro** [comment]
`code_sniffer` is great!

2 annoying rules though:
  * 85 char per line is not enough.
  * Requiring `@return void` is useless.

File: application/Config.php (hunk 4 above)
Comment (virtualtam) [inline application/Config.php]: it's a good practice: this way, a behavior is not altered by external changes

a future evolution could be to handle config entries within a proper class, and make this list static

**Comment by virtualtam** [comment]
> is there an issue where this is discussed?

Yep, in #95 :)

> annoying rules

there are plenty of them, the tools' parameters need to be refined to fit our needs:
* line length: 80 as a recommendation, 120 as a hard limit
* PHP headers (author info, license...): as the sources are versioned, they are of little use
* conditional structures: Shaarli may need to undergo a bit of refactoring to simplify them :p

the features I found the most useful are:
* indents
* variable & method naming
* access to globals
* unused objects
* cyclomatic (aka McCabe's) complexity

File: application/Config.php (hunk 4 above)
Comment (ArthurHoaro) [inline application/Config.php]: Alright. I'll declare the list within `writeConfig` for now, but I agree, a class might be a better choice.

**Comment by virtualtam** [comment]
+1, works fine :)

Could you squash the commits into a single one before merging?

**Comment by ArthurHoaro** [comment]
Should be fine.

**Pull request merged**